Mempool design (#997)
input-output-hk/ouroboros-leios: docs/leios-design/README.md

2. **Task scheduler** that is managing transaction **validation** and their **application** to one or more ledger states. This distinction is crucial in the Mempool and leveraged in the existing implementation, arguably not to its fullest potential.  **Transaction validation** (through `applyTx`) has to **only ever be performed once, can be cached and performed independently from state** (with some caveats), which is great because that is the most expensive part. **Transaction application** (through `reapplyTx`) however is inherently sequential (also with some caveats) and has to be performed on each (base) state changes.

#### Transaction processing modes

The distinction between the two modes, **validation** and **application**, can be loosely depicted using jigsaw puzzles. Validating a puzzle piece means checking its features are aligned with the overall puzzle (brand, color, shape, thickness, material), applying a puzzle piece means finding whether it fits somewhere in the currently assembled state.

Bringing it back to transactions...

> Validating a transactions means checking signatures, scripts, limits, values etc. Applying a transaction however, means checking whether the state it builds upon to produce new state exists (utxos, stake pool, reward accounts, proposals, **time** etc).

Reviewer: This is statement is too loose in my opinion. Validation of a transaction is not just checking signatures and limits, it is also ensuring that transaction is valid against the current ledger state, eg. withdrawal is not taking out more funds from the account that is available in it or utxo being spent is actually present in the state.

That being said, if you think of transaction that is being "applicable" is the one that is "valid" against the existing ledger state, I'd suggest changing this terminology in this document, because that does not align with the terminology that we have today.

Terminology of today is:
* "static" validation is checking all of the transaction parts that doe not depend on the state
* "non-static" validation is checking all parts of the transaction that do depend on the state

I believe this is more descriptive, because we also have a mode where simply apply transaction without validation, which can be easily confused with as "applicable"

To hammer this down it's useful to say that a transaction can be **valid** and **applicable**. If a transaction is "valid" it can (but doesn't have to) be "applicable". On the other hand, if a transaction is "applicable" then it can (but doesn't have to) be "valid". We usually want both, but it's useful to know how they relate as it will help us define what we want to diffuse (valid? valid and applicable?) and how to organize tasks in the Mempool for optimal performance.

Q: What about UTxOs during validation, don't you need to first fetch them from storage, isn't that necessarily state dependent? What if you have dependent transactions? Doesn't that dissolve the distinction with "application" and makes "validation" inherently sequential?
A: Yes and no!

  - Yes, in that UTxOs form part of the state in Cardano and need to be **resolved** before we can even start validating transactions either by querying storage or preceding transaction outputs. In other words, transactions in Cardano are **incomplete** and need a separate **resolution** step before validation that makes them **complete**.

>   - No, in that if a transaction validates it's valid **forever** (actually for an era), even though state it refers to might have been changed and made it **inapplicable** to some particular state. How one **resolves** transactions and makes them **complete** can be sequential, but doesn't have to.

Reviewer: This is a wild overstatement.
* It is possible for a plutus script to change its outcome after an intra-era hard fork. So even a static check, can change its behavior within an era. This is just one example, but we can have more, because that is the point of hard fork: change the validity of a transaction
* Transactions naturally become valid or invalid for a specific slotNo
* Transactions can conflict or depend on each other in many different ways, which is can only be proven through the ledger state. Albeit inter-dependency can be inferred simply through looking at any two transactions.

Let's attempt such a model that can make it simple to pose useful questions against the current and future implementations.

```haskell
type LedgerState = RbHash -- Ledger state is uniquely identified by the Praos (ranking) block hash

data MempoolAction = AddTx Tx -- peers and local clients (1000 per s)
  | Rebase LedgerState -- chain selection (several per 20s)
  | Tick SlotNo -- time passing (1 per s)

-- Mempool state holds the `base` ledger state,
-- the `txs` that were resolved, validated and
-- applicable to `base` ledger state AND `slot`.
data MempoolState = MempoolState {
  slot :: SlotNo,
  base :: LedgerState,
  txs : [Tx],
}
```

The main questions we need to answer:

1. How do we decide which transactions to accept and diffuse?
2. How do we decide which transactions are eligible to be included in the new block?

##### `Tick`

`slot` is a state variable that changes on `Tick` action and can affect both transaction "validity" and "applicability".

(In Cardano, time is denoted with slots, and is handled separate to `LedgerState`, not strictly necessary but it's worth mirroring that implementation choice.)

> It can affect **validity** because slots passing through epoch boundaries can change protocol parameters and the "ledger law". Since validity is essentially "ledger law" judgement on transactions, a transaction that was valid in one epoch is not necessarily valid in another. In other words, validity is defined per "ledger era". For example, script limits could change between eras.

Reviewer: There is also an explicit validity interval field in a Tx. So it does affect validity directly.
I guess you probably mean "applicability" hear. But that is precisely why I have a problem with this "new" terminology. Even the name of the field literally says "validityInterval"

```suggestion
It can affect **validity** because slots passing through epoch boundaries can change protocol parameters and the "ledger law". Since validity is essentially "ledger law" judgement on transactions, a transaction that was valid in one epoch is not necessarily valid in another. In other words, validity is defined per "ledger era". For example, script limits could change between protocol versions.
```

> (Q: Is it the case that if a Cardano transaction is valid in era N it's valid in all future M > N eras? Given the transaction "forward translation" feature? If so, we don't have to re-validate transactions when crossing epochs/eras?)

Reviewer: No it is not.
Forward translation feature does not guarantee translation, nor does it guarantee validity.
```suggestion
```

> It can also affect **applicability** because the slot can be outside of transaction interval. Another way applicability can change is, again, when passing through epoch boundaries as certain actions are put in effect late (pool registration etc.) and some state variables are created or deleted (stake pools). For example, a transaction with "always" interval that retires a pool A can be applicable in one epoch but not another.

Reviewer 2: Also protocol parameters can change on epoch boundaries, so (for example) a deposit for registration which was valid in one epoch may not be in another.  This puts protocol parameters into the "applicability" context, I think.

> (In the current terminology it would make sense to say transaction applicability interval, rather than validity interval)

Reviewer: As you well spotted it yourself 😄 
So, instead of changing terminology after half a decade of developers and users around the World using it, it would be better to stick to the same terminology as we have today: static vs non-static validity.
Granted, what we have today is not a commonly known terminology, so we could change that, but changing the notion of transaction "validity" as we know it today is unacceptable. 
```suggestion
```

That being said, time also affects the ledger state itself, as some state variables are "internal" to it and defined in terms of time, and not observable by transactions (for example incremental rewards accounting). In Cardano, such effects are accumulated incrementally and "published" on epoch boundaries.

Why do we care about ticking in the Mempool? Well, that's the same as answering our two main questions wrt time...

1. How do we decide which transactions to accept and diffuse?
  - Transactions that are valid and applicable to `slot` are accepted and diffused.
2. How do we decide which transactions are eligible to be included in the new block?
  - Transactions that are valid and applicable to `slot` which matches the election winning slot can be included verbatim in the new block.

Scenario: Bob is sending a transaction X to Alice, but they don't know what `slot` the others' Mempool has.

How will Alice determine to accept or reject the transaction X (and potentially punish Bob)?
After all, there's no guarantee both Mempool `slot`s match!

> Perhaps in Bob's Mempool transaction X is valid and applicable, just on a different `slot`.

Reviewer: It is not only the slot, but also the set of transactions that Alice has can determine whether to accept Bobs transaction, there are no guarantees that they will have the same set.
Even the order in which they received transaction could affect the set of transactions they both will have in their mempool

And there is no protocol currently with which they can share that information.
Since Bob and Alice are currently just guessing, they can't determine whether some behavior is adversarial.

(TODO: Include Alexey's epoch tick measurement data)

Let's examine the ticking pseudocode:

1. on `Tick newSlot` (newSlot > slot)
  1. update `slot` to `newSlot`
  2. tick `base` to `newSlot`
    1. update internal state variables
    2. if `newSlot` in new epoch
      1. update state variables observable to `txs`
  3. if `newSlot` in new epoch
    1. if `newSlot` in new era
      2. re-validate `txs`
  4. re-apply `txs`

> Are we really going to do this every second?

Reviewer: Ticks today happen on every slot that has a block, not every second.

Step `1.2` can be fast, but on epoch boundary it's actually quite slow.

> Step `1.3.2` is always expensive, but re-validation only has to happen on new era.

Reviewer: ```suggestion
Step `1.3.1.2` is always expensive, but re-validation only has to happen on new protocol version.
```

Step `1.4` can be made fast by using "smarter" scheduling techniques, but for now is also slow.

Is there something better we can do?

I propose the **Mempool transaction interval** that rejects transactions with interval that don't contain it.
This intervals' left bound is the maximum between tip slot and last epoch boundary, and the right bound is minimum between "expected next block slot** and next epoch boundary.
In practice, this interval would be ~20 slots long, arguably an acceptable tradeoff that gives correctness and performance but takes away transactions with an interval smaller than 20 slots.

> (TODO: Include Kosta's mainnet interval analysis)

Reviewer: 😄 
```suggestion
(TODO: Include Kostas' mainnet interval analysis)
```

    2. re-apply `txs` (re-validate if new era)
  2. if `newSlot` outside `interval` and not in new epoch (TODO: We should have seen a new block by now and rebase?)
  3. if `newSlot` inside `interval`
    1. `tick` internal `base` state variables

And back to the main questions:

1. How do we decide which transactions to accept and diffuse?
  - Transactions that are valid and applicable to `interval` are accepted and diffused.
2. How do we decide which transactions are eligible to be included in the new block?
  - Transactions that are valid and applicable to `interval` which contains the election winning slot can be included verbatim in the new block.

##### `Rebase`

`base` is a state variable that should reflect the currently selected ledger state. When a selection changes, Mempool needs to rebase `txs` on top of this new `base` (and `interval`).

Since there can be many transaction in `txs` it can take a long time to re-apply this sequence.
It's noteworthy to reiterate that we only need to **apply** and not **validate** `txs` as we established before that **validation** can be cached forever (or for-era) while **application** is state dependent.

(TODO: Insert some data here)

> (The very fact that we treat them as a sequence is unfortunate, as especially in Cardano ledger there's many opportunities to schedule this work more intelligently by treating transactions as a DAG by utilizing transaction topology. However, that's currently a research and development question that requires a non-trivial amount of effort.)

Reviewer 2: I looked at topological dependency ordering in Acropolis and found for the dominant case of UTXOs (there are other dependencies - e.g. registrations), because the validation (in your terms, independent of order) can be done in advance and in parallel, by the time you come to do the order-dependent stuff you may as well keep it linear, particularly as the state you're modifying (the UTXO set) has to be mutexed anyway.

  3. re-apply `txs`

Scenario: Bob is sending Alice 100 txs per second, and Alice is going through a rebase, what happens?

In the current implementation, we're utilizing a wait-lock, during a rebase all access to the Mempool is blocked until the rebase is done.
If a rebase takes 1 second, Bob will queue up 100 transactions waiting on the rebase to finish, and in that time Bob's transactions aren't diffused.
In a naive, lock-free solution, Bob might be able to add transactions to Alice's Mempool while the rebase is ongoing and diffusion would continue,
but when the rebase is done and its work needs to be committed, we're faced with a similar problem:
there's 100 transactions more in the Mempool than when the rebase started. What do we do with them? Drop them? Or re-apply them too?
If we re-apply them, do we do that in some priority way? Perhaps with a lock?

Sounds like this is an improvement! We're not waiting on a longer rebase but instead waiting on a shorter "merge" that reconciliates
conflicts due to Mempool effects occurring concurrently (namely re-applying 100 surplus transactions).

```haskell

mergeMempoolState :: MempoolState -> MempoolState -> MempoolState

> mergeMempoolState lm rm  = -- in the case of miss matching `base`, 'select' the one with the "better" base and

Reviewer: It is theoretically impossible to merge two separate ledger states, unless there share a common base and effects of transactions are known and transactions are disjoint in their effects.
For example, imagine there is a base ledger state with account `X` that has 10ADA in it. Then `lr` has a transactions that deposits 5 ADA into account `X`, while `rm` has a transaction that withdraws 10ADA and unregisters account `X`. Clearly `rm` is in conflict with `lr`

We have to resolve transactions in sequence, in case transactions are dependent, the state they produce (UTxOs) might be needed by some following transactions.

Therefore, given a sequence of transactions, we can do a resolution pass to make them complete.
This sounds enticing, but what if those transactions turn out to be invalid? Or valid but inapplicable?

One goal of the Mempool is to minimize work that an adversary can make it do "for free".

Until a topology aware transaction scheduler comes about, we're seemingly stuck with processing every transaction in sequence: resolve, validate and apply.

#### TODO: EB extended `base`

#### Final model, concurrency and protocol changes

Given the final proposed model

```haskell
type LedgerState = RbHash -- Ledger state is uniquely identified by the Praos (ranking) block hash

data MempoolAction = AddTx Tx -- peers and local clients (1000 per s)
  | Rebase LedgerState -- chain selection (several per 20s)
  | Tick SlotNo -- time passing (1 per s)

-- Mempool state holds the `base` ledger state,
-- the `txs` that were resolved, validated and
-- applicable to `base` ledger state AND `interval`.
data MempoolState = MempoolState {

>   interval :: (SlotNo, SlotNo),

Reviewer: What happens with a transaction that has a validity interval that has only a partial intersection with this interval? The only correct thing to do would be to exclude such transaction, but that would mean this would artificially reduce number of transactions that would actually make it on chain.
